barman/server.py

```diff
@@ -1515,6 +1515,65 @@ def show(self):
     def delete_backup(self, backup):
         """Deletes a backup
 
+        :param backup: the backup to delete
+        """
+        if self.backup_manager.should_keep_backup(backup.backup_id):
+            output.warning(
+                "Skipping delete of backup %s for server %s "
+                "as it has a current keep request. If you really "
+                "want to delete this backup please remove the keep "
+                "and try again.",
+                backup.backup_id,
+                self.config.name,
+            )
+            return
+
+        # Honour minimum required redundancy
+        available_backups = self.get_available_backups(status_filter=(BackupInfo.DONE,))
+        minimum_redundancy = self.config.minimum_redundancy
+        if backup.status == BackupInfo.DONE and minimum_redundancy >= len(
+            available_backups
+        ):
+            output.warning(
+                "Skipping delete of backup %s for server %s "
+                "due to minimum redundancy requirements "
+                "(minimum redundancy = %s, "
+                "current redundancy = %s)",
+                backup.backup_id,
+                self.config.name,
+                minimum_redundancy,
+                len(available_backups),
+            )
+            return
+
+        if backup.children_backup_ids:
+            output.warning(
+                "Backup %s has incremental backups which depent on it. "
+                "Deleting all backups in the tree",
+                backup.backup_id,
+            )
+
+            # Delete the backup along with all its descendants in the
+            # backup tree i.e. all its subsequent incremental backups.
+            # If it has no descendants, only the current backup is deleted.
+            deleted = False
+            backups_to_delete = backup.walk_backups_tree()
+            for del_backup in backups_to_delete:
+                deleted = self.perform_delete_backup(del_backup)
+                if not deleted and del_backup.backup_id != backup.backup_id:
+                    output.error(
+                        "Failed to delete an incremental backup of the specified backup. "
+                        "Make sure all its dependent backups are deletable and try again."
+                    )
+                    break
+        else:
+            deleted = self.perform_delete_backup(backup)
+
+        return deleted
+
+    def perform_delete_backup(self, backup):
+        """Deletes a backup
```

Comment:
I'd say to add `:returns:` to both `delete_backup` and `perform_delete_backup`, only for docstrings completeness.